File: backend/src/main/java/com/productivity/service/TaskService.java
package com.productivity.service;

import com.google.cloud.Timestamp;
import com.productivity.dto.SubtaskDTO;
import com.productivity.dto.TaskDTO;
import com.productivity.exception.ResourceNotFoundException;
import com.productivity.model.Task;
import com.productivity.repository.TaskRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

@Service
public class TaskService {

    private static final Logger log = LoggerFactory.getLogger(TaskService.class);
    private final TaskRepository taskRepository;

    public TaskService(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    public TaskDTO createTask(String userId, TaskDTO dto) throws ExecutionException, InterruptedException {
        Task task = new Task();
        task.setTitle(dto.getTitle());
        task.setDescription(dto.getDescription());
        task.setPriority(dto.getPriority() != null ? dto.getPriority() : "medium");
        task.setStatus(dto.getStatus() != null ? dto.getStatus() : "todo");
        task.setCreatedAt(Timestamp.now());
        task.setUpdatedAt(Timestamp.now());

        if (dto.getDeadline() != null) {
            task.setDeadline(Timestamp.ofTimeSecondsAndNanos(
                    Instant.parse(dto.getDeadline()).getEpochSecond(), 0));
        }

        if (dto.getSubtasks() != null) {
            task.setSubtasks(dto.getSubtasks().stream().map(s -> {
                Map<String, Object> map = new HashMap<>();
                map.put("id", s.getId());
                map.put("title", s.getTitle());
                map.put("completed", s.isCompleted());
                return map;
            }).collect(Collectors.toList()));
        }

        Task saved = taskRepository.save(userId, task);
        return toDTO(saved);
    }

    public TaskDTO getTask(String userId, String taskId) throws ExecutionException, InterruptedException {
        Task task = taskRepository.findById(userId, taskId)
                .orElseThrow(() -> new ResourceNotFoundException("Task", taskId));
        return toDTO(task);
    }

    public List<TaskDTO> getTasks(String userId, String status, String priority, String sortBy,
                                   String sortDirection, int page, int size) throws ExecutionException, InterruptedException {
        List<Task> tasks = taskRepository.findAll(userId, status, priority, sortBy, sortDirection, page, size);
        return tasks.stream().map(this::toDTO).collect(Collectors.toList());
    }

    public long getTaskCount(String userId) throws ExecutionException, InterruptedException {
        return taskRepository.count(userId);
    }

    public TaskDTO updateTask(String userId, String taskId, TaskDTO dto) throws ExecutionException, InterruptedException {
        Task existing = taskRepository.findById(userId, taskId)
                .orElseThrow(() -> new ResourceNotFoundException("Task", taskId));

        if (dto.getTitle() != null) existing.setTitle(dto.getTitle());
        if (dto.getDescription() != null) existing.setDescription(dto.getDescription());
        if (dto.getPriority() != null) existing.setPriority(dto.getPriority());
        if (dto.getStatus() != null) existing.setStatus(dto.getStatus());
        if (dto.getDeadline() != null) {
            existing.setDeadline(Timestamp.ofTimeSecondsAndNanos(
                    Instant.parse(dto.getDeadline()).getEpochSecond(), 0));
        }

        if (dto.getSubtasks() != null) {
            existing.setSubtasks(dto.getSubtasks().stream().map(s -> {
                Map<String, Object> map = new HashMap<>();
                map.put("id", s.getId());
                map.put("title", s.getTitle());
                map.put("completed", s.isCompleted());
                return map;
            }).collect(Collectors.toList()));
        }

        existing.setUpdatedAt(Timestamp.now());

        Task saved = taskRepository.save(userId, existing);
        return toDTO(saved);
    }

    public void deleteTask(String userId, String taskId) throws ExecutionException, InterruptedException {
        taskRepository.findById(userId, taskId)
                .orElseThrow(() -> new ResourceNotFoundException("Task", taskId));
        taskRepository.delete(userId, taskId);
    }

    public List<TaskDTO> getOverdueTasks(String userId) throws ExecutionException, InterruptedException {
        return taskRepository.findOverdue(userId, Timestamp.now())
                .stream().map(this::toDTO).collect(Collectors.toList());
    }

    public List<TaskDTO> getTodayTasks(String userId) throws ExecutionException, InterruptedException {
        LocalDate today = LocalDate.now(ZoneOffset.UTC);
        Timestamp startOfDay = Timestamp.ofTimeSecondsAndNanos(
                today.atStartOfDay(ZoneOffset.UTC).toEpochSecond(), 0);
        Timestamp endOfDay = Timestamp.ofTimeSecondsAndNanos(
                today.plusDays(1).atStartOfDay(ZoneOffset.UTC).toEpochSecond(), 0);
        return taskRepository.findByDeadlineRange(userId, startOfDay, endOfDay)
                .stream().map(this::toDTO).collect(Collectors.toList());
    }

    private TaskDTO toDTO(Task task) {
        TaskDTO dto = new TaskDTO();
        dto.setId(task.getId());
        dto.setTitle(task.getTitle());
        dto.setDescription(task.getDescription());
        dto.setPriority(task.getPriority());
        dto.setStatus(task.getStatus());
        if (task.getDeadline() != null) {
            dto.setDeadline(task.getDeadline().toDate().toInstant().toString());
        }

        if (task.getSubtasks() != null) {
            dto.setSubtasks(task.getSubtasks().stream().map(s -> {
                SubtaskDTO sub = new SubtaskDTO();
                sub.setId((String) s.get("id"));
                sub.setTitle((String) s.get("title"));
                sub.setCompleted((Boolean) s.get("completed"));
                return sub;
            }).collect(Collectors.toList()));
        }

        if (task.getCreatedAt() != null) {
            dto.setCreatedAt(task.getCreatedAt().toDate().toInstant().toString());
        }
        if (task.getUpdatedAt() != null) {
            dto.setUpdatedAt(task.getUpdatedAt().toDate().toInstant().toString());
        }
        return dto;
    }
}

File: backend-node/src/services/taskService.js
'use strict';

const { v4: uuidv4 } = require('uuid');
const taskRepo = require('../repositories/taskRepository');
const { toISO } = require('../utils/timestamp');

function notFound(taskId) {
  return Object.assign(new Error(`Task not found: ${taskId}`), { status: 404 });
}

async function createTask(userId, dto) {
  const task = {
    title: dto.title,
    description: dto.description || null,
    priority: dto.priority || 'medium',
    status: dto.status || 'todo',
    subtasks: dto.subtasks ? dto.subtasks.map(s => ({
      id: s.id || uuidv4(),
      title: s.title,
      completed: s.completed || false,
    })) : [],
    externalLinks: dto.externalLinks || [],
    deadline: dto.deadline ? new Date(dto.deadline) : null,
    createdAt: new Date(),
    updatedAt: new Date(),
  };

  const saved = await taskRepo.save(userId, task);
  return toDTO(saved);
}

async function getTask(userId, taskId) {
  const task = await taskRepo.findById(userId, taskId);
  if (!task) throw notFound(taskId);
  return toDTO(task);
}

async function getTasks(userId, { status, priority, sortBy, sortDirection, page, size } = {}) {
  const tasks = await taskRepo.findAll(userId, { status, priority, sortBy, sortDirection, page, size });
  return tasks.map(toDTO);
}

async function getTaskCount(userId) {
  return taskRepo.count(userId);
}

async function updateTask(userId, taskId, dto) {
  const existing = await taskRepo.findById(userId, taskId);
  if (!existing) throw notFound(taskId);

  if (dto.title != null) existing.title = dto.title;
  if (dto.description != null) existing.description = dto.description;
  if (dto.priority != null) existing.priority = dto.priority;
  if (dto.status != null) existing.status = dto.status;
  if (dto.deadline != null) existing.deadline = new Date(dto.deadline);
  if (dto.subtasks != null) {
    existing.subtasks = dto.subtasks.map(s => ({
      id: s.id || uuidv4(),
      title: s.title,
      completed: s.completed || false,
    }));
  }
  if (dto.externalLinks != null) existing.externalLinks = dto.externalLinks;
  existing.updatedAt = new Date();

  const saved = await taskRepo.save(userId, existing);
  return toDTO(saved);
}

async function deleteTask(userId, taskId) {
  const task = await taskRepo.findById(userId, taskId);
  if (!task) throw notFound(taskId);
  await taskRepo.remove(userId, taskId);
}

async function getOverdueTasks(userId) {
  const tasks = await taskRepo.findOverdue(userId);
  return tasks.map(toDTO);
}

async function getTodayTasks(userId) {
  const now = new Date();
  const startOfDay = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()));
  const endOfDay = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate() + 1));
  const tasks = await taskRepo.findByDeadlineRange(userId, startOfDay, endOfDay);
  return tasks.map(toDTO);
}

function toDTO(task) {
  return {
    id: task.id,
    title: task.title,
    description: task.description || null,
    priority: task.priority,
    status: task.status,
    deadline: toISO(task.deadline),
    subtasks: task.subtasks || [],
    externalLinks: task.externalLinks || [],
    createdAt: toISO(task.createdAt),
    updatedAt: toISO(task.updatedAt),
  };
}

module.exports = { createTask, getTask, getTasks, getTaskCount, updateTask, deleteTask, getOverdueTasks, getTodayTasks };
